### PR TITLE
Add `seed` parameter to make initial configuration sampling deterministic

### LINKED ISSUE #112: [Feature Request] Reproducible results for TuneSearchCV
Is it possible to produce reproducible results for TuneSearchCV using a random seed?

### PR DESCRIPTION
**Pull request by krfricke**
For reproducible results, we need to be able to set a random seed. This PR introduces the respective argument to `TuneSearchCV`. 

Please note that this mostly affects initial configuration sampling. With parallel evaluations, there's always non-determinism involved that can lead to different results of the training run.

Currently not implemented on BOHB.

Closes #112

### CODE REVIEW

**Comment by Yard1** [comment]
FYI test for BOHB is failing. Aside from that, looks good. One idea for better sklearn compatibility - do you think it would be possible to pass numpy `RandomState` instances in addition to integers - https://scikit-learn.org/stable/glossary.html#term-random-state? It should be possible to just get an integer seed from those and then just carry on with an integer.

**Comment by krfricke** [comment]
This is weird, it works for me locally every time I run it, even when I alter the search space. I'll look into this tomorrow.
I didn't know about `RandomState` but will make sure that it is supported. Thanks!

**Comment by krfricke** [comment]
Before I get to this, should we just use the existing `random_state` parameter for the other search algorithms?

**Comment by Yard1** [comment]
I think that would be the best, honestly. I forgot that it's already there! I think a user would expect just one randomness parameter to control everything. I believe that's also how other libraries do it. I don't see much utility in splitting it into `random_state` and `seed`.

File: tune_sklearn/tune_search.py [hunk 1]
```diff
@@ -262,6 +263,9 @@ class TuneSearchCV(TuneBaseSearchCV):
             seconds after which all trials are stopped. Can also be a
             ``datetime.timedelta`` object. The stopping condition is checked
             after receiving a result, i.e. after each training iteration.
+        seed (int): Seed used to initialize random number generators. Running
```
Comment (Yard1) [inline tune_sklearn/tune_search.py:266]: Leftover docstring

**Review by Yard1 — CHANGES_REQUESTED** [review]
Just one leftover docstring, rest looks perfect

**Comment by krfricke** [comment]
But the tests seem to fail nevertheless. Weird. Could you run them locally and see if you get the same errors? I don't have any errors on my machine, no matter how often I run them.

**Comment by krfricke** [comment]
Ah, so it's still just BOHB. Maybe there's a dependency mismatch or so. I'll see what I can find.

**Comment by krfricke** [comment]
So even on a fresh env with py 3.6.10 it passes for me. I tried 10 different seeds sequentially, all passed. Any ideas? 
The only difference I see right now is that I'm on mac and the VM is on linux. I could try setting up a new env on a cluster machine later or try on my linux laptop.

**Comment by Yard1** [comment]
I wouldn't be surprised if it was something to do with the OS. ConfigSpace uses Cython, so I'd imagine it's not as OS agnostic as other libraries. In the worst case, we may have to just not test for it. One thing we can try, however, is setting `n_jobs=1`, so it's ran in Local Mode. Perhaps that may be it?

**Review by Yard1 — APPROVED** [review]

**Comment by richardliaw** [comment]
Hmm, looks like the seed test is failing on mac os py3.7?

**Comment by Yard1** [comment]
It's BOHB again D:

**Comment by krfricke** [comment]
Ok I think I got it - search space conversions might disregard custom seed. This PR in ray should fix this by introducing a `seed` parameter for BOHB: https://github.com/ray-project/ray/pull/12160

After that one is merged, tests with latest ray should pass. Let's defer this PR until then.

**Comment by krfricke** [comment]
It was merged, I'll wait for the next nightly wheels, push an update to this PR and then fingers crossed.

File: tune_sklearn/tune_search.py [hunk 2]
```diff
@@ -204,10 +206,9 @@ class TuneSearchCV(TuneBaseSearchCV):
             distributions.
             If int, random_state is the seed used by the random number
             generator;
-            If RandomState instance, random_state is the random number
-            generator;
+            If RandomState instance, a seed is sampled from random_state;
             If None, the random number generator is the RandomState instance
-            used by np.random. Defaults to None.
+            used by np.random and no seed is passed. Defaults to None.
```
Comment (richardliaw) [inline tune_sklearn/tune_search.py:211]:
```suggestion
            used by np.random and no seed is provided. Defaults to None.
```
